stardis/plasma/base.py

```diff
@@ -129,6 +129,145 @@ def calculate(
+        ].merge(
+            g.loc[:, :, 0].rename("g_0"),
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        # Truncate to final atomic number
+        linelist = linelist[
+            linelist.atomic_number < (atomic_data.selected_atomic_numbers.max() + 1)
+        ]
+
+        # Calculate degeneracies
+        linelist["g_lo"] = linelist.j_lo * 2 + 1
+        linelist["g_up"] = linelist.j_up * 2 + 1
+        linelist["g"] = linelist.g_lo / linelist.g_0
+
+        exponent_by_point = np.exp(
+            np.outer(
+                -linelist.e_low.values * u.eV, 1 / (t_electrons * u.K * const.k_B)
+            ).to(1)
+        )
+
+        # grab densities for n_lower - need to use linelist as the index
+        linelist_with_densities = linelist.merge(
+            ion_number_density,
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        n_lower = (
+            exponent_by_point * linelist_with_densities[np.arange(points)]
+        ).values.T * linelist_with_densities.g.values
+
+        linelist["f_lu"] = 10**linelist.log_gf * linelist.g_up / linelist.g_lo
+
+        line_nus = (linelist.wavelength.values * u.AA).to(
+            u.Hz, equivalencies=u.spectral()
+        )
+
+        emission_correction = 1 - np.exp(
+            (
+                -const.h
+                / const.k_B
+                * np.outer(
+                    line_nus,
+                    1 / (t_electrons * u.K),
+                )
+            ).to(1)
+        )
+
+        alpha = (
+            ALPHA_COEFFICIENT * n_lower * linelist.f_lu.values * emission_correction.T
+        )
+
+        ###TODO - include warning for missing values.
+        # if np.any(np.isnan(alpha)) or np.any(np.isinf(np.abs(alpha))):
+        #     raise ValueError(
+        #         "Some alpha_line are nan, inf, -inf " " Something went wrong!"
+        #     )
+
+        df = pd.DataFrame(
+            alpha.T,
+        )
+
+        df["nu"] = line_nus.value
+        linelist["nu"] = line_nus.value
+
+        # Linelist preparation below is taken from opacities_solvers/base/calc_alpha_line_at_nu
+        ionization_energies = ionization_data.reset_index()
+        ionization_energies["ion_number"] -= 1
+        linelist = pd.merge(
+            linelist,
+            ionization_energies,
+            how="left",
+            on=["atomic_number", "ion_number"],
+        )
+
+        linelist["level_energy_lower"] = ((linelist["e_low"].values * u.eV).cgs).value
+        linelist["level_energy_upper"] = ((linelist["e_up"].values * u.eV).cgs).value
+        ###TODO THIS IS WRONG - Fix this later
+        linelist["A_ul"] = 10 ** (
+            linelist["rad"]
+        )  # see 1995A&AS..112..525P for appropriate units - may be off by a factor of 4pi
+
+        # this might be improperly named - it's actually the indices of everything that aren't autoionizing
+        autoionization_indices = (
```

Comment:
Definitely not a good name. I would suggest "valid_indices" perhaps? This would ideally be temporary pending the quantum number calculation.